File: Caronae/User/TokenViewController.m
#import <SafariServices/SafariServices.h>
#import <SVProgressHUD/SVProgressHUD.h>
#import "AppDelegate.h"
#import "CaronaeAlertController.h"
#import "EditProfileViewController.h"
#import "NSDictionary+dictionaryWithoutNulls.h"
#import "TokenViewController.h"
#import "Caronae-Swift.h"

@interface TokenViewController () <UITextFieldDelegate, CaronaeSignInDelegate>

@end

@implementation TokenViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    _authButton.enabled = NO;
    _tokenTextField.delegate = self;
    _idTextField.delegate = self;
    
    UITapGestureRecognizer *welcomeTextTapRecognizer = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(didTapWelcomeText:)];
    [_welcomeLabel addGestureRecognizer:welcomeTextTapRecognizer];
    
    dispatch_async(dispatch_get_main_queue(), ^{
        [_createUserButton removeFromSuperview];
    });
}

- (void)authenticateWithUser:(NSString *)user token:(NSString *)token {
    _authButton.enabled = NO;
    [self.view endEditing:YES];
    [SVProgressHUD show];
    
    NSDictionary *params = @{ @"id_ufrj": user,
                              @"token": token };
    [CaronaeAPIHTTPSessionManager.instance POST:@"/user/login" parameters:params success:^(NSURLSessionDataTask * _Nonnull task, id  _Nonnull responseObject) {
        [SVProgressHUD dismiss];
        
        // Check if the authentication was ok if we received an user object
        if (responseObject[@"user"]) {
            // Save user's created rides
            NSArray *rides = responseObject[@"rides"];
            NSMutableArray *filteredRides = [NSMutableArray arrayWithCapacity:rides.count];
            for (id rideDictionary in rides) {
                [filteredRides addObject:[rideDictionary dictionaryWithoutNulls]];
            }
            
            NSError *error;
            User *user = [MTLJSONAdapter modelOfClass:User.class fromJSONDictionary:responseObject[@"user"] error:&error];
            if (error) {
                [CaronaeAlertController presentOkAlertWithTitle:@"Não foi possível autenticar" message:@"Ocorreu um erro carregando seu perfil."];
                _authButton.enabled = YES;
                return;
            }
            
            [[UserController sharedInstance] setUser:user token:token rides:filteredRides];
            
            AppDelegate *appDelegate = (AppDelegate*) [UIApplication sharedApplication].delegate;
            [appDelegate registerForNotifications];
            
            [self performSegueWithIdentifier:@"ViewHome" sender:self];
        }
        else {
            NSLog(@"Error authenticating");
            _authButton.enabled = YES;
        }
    } failure:^(NSURLSessionDataTask * _Nullable task, NSError * _Nonnull error) {
        [SVProgressHUD dismiss];
        NSLog(@"Error trying to authenticate: %@", error.localizedDescription);
        
        NSString *errorMsg;
        NSHTTPURLResponse *response = (NSHTTPURLResponse*)task.response;
        if (response.statusCode == 403) {
            errorMsg = @"Chave não autorizada. Verifique se a mesma foi digitada corretamente e tente de novo.";
        }
        else {
            errorMsg = [NSString stringWithFormat:@"Ocorreu um erro autenticando com o servidor do Caronaê. Tente novamente.\n(%@)", error.localizedDescription];
        }
        
        [CaronaeAlertController presentOkAlertWithTitle:@"Não foi possível autenticar" message:errorMsg];
        _authButton.enabled = YES;
    }];
}


#pragma mark - Navigation

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    if ([segue.identifier isEqualToString:@"CompleteProfile"]) {
        UINavigationController *editProfileNavController = segue.destinationViewController;
        EditProfileViewController *vc = editProfileNavController.viewControllers.firstObject;
        vc.completeProfileMode = YES;
    }
}

#pragma mark - IBActions

- (IBAction)didTapAuthenticateButton:(UIButton *)sender {
    NSString *userToken = _tokenTextField.text;
    NSString *idUFRJ = _idTextField.text;
    [self authenticateWithUser:idUFRJ token:userToken];
}

- (void)didTapWelcomeText:(id)sender {
    [CaronaeSignInViewController presentFromViewController:self delegate:self];
}


#pragma mark - Sign in delegate

- (void)caronaeDidSignInWithSuccessWithUser:(NSString *)user token:(NSString *)token {
    [self authenticateWithUser:user token:token];
}

- (void)caronaeSignInFailed {
    NSLog(@"Authentication failed");
}

#pragma mark Text field methods

- (BOOL)textFieldShouldReturn:(UITextField *)textField {
    if (textField == _tokenTextField) {
        NSString *userToken = _tokenTextField.text;
        NSString *idUFRJ = _idTextField.text;
        [self authenticateWithUser:idUFRJ token:userToken];
        return NO;
    }
    else if (textField == _idTextField && _idTextField.hasText) {
        [_tokenTextField becomeFirstResponder];
    }
    
    return YES;
}

- (BOOL)textField:(UITextField *)textField shouldChangeCharactersInRange:(NSRange)range replacementString:(NSString *)string {
    NSString *text = [textField.text stringByReplacingCharactersInRange:range withString:string];
    if (textField == _idTextField) {
        _authButton.enabled = ![text isEqualToString:@""] && _tokenTextField.hasText;
    }
    else if (textField == _tokenTextField) {
        _authButton.enabled = ![text isEqualToString:@""] && _idTextField.hasText;
    }
    
    return YES;
}


@end

File: Caronae/User/TokenViewController.swift
import SafariServices
import SVProgressHUD
import UIKit

class TokenViewController: UIViewController {
    @IBOutlet weak var welcomeLabel: UILabel!
    @IBOutlet weak var authButton: UIButton!
    @IBOutlet weak var idTextField: UITextField!
    @IBOutlet weak var tokenTextField: UITextField!
    var authController: AuthenticationController!
    
    static func tokenViewController() -> TokenViewController
    {
        let viewController = UIStoryboard(name: "Main", bundle: nil).instantiateViewController(withIdentifier: "InitialTokenScreen") as! TokenViewController
        viewController.modalTransitionStyle = .flipHorizontal
        return viewController
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        
        if #available(iOS 11, *) {
            self.authController = AuthenticationController()
            self.authController.authenticate(success: { (idUFRJ: String, token: String) in
                NSLog("Authenticated with id '%@' and token '%@'", idUFRJ, token)
                self.authenticate(id: idUFRJ, token: token)
            })
        }

        authButton.isEnabled = false
        tokenTextField.delegate = self
        idTextField.delegate = self
        
        let tapRecognizer = UITapGestureRecognizer(target: self, action: #selector(openIntranetLogin))
        welcomeLabel.addGestureRecognizer(tapRecognizer)
    }

    func authenticate(id: String, token: String) {
        authButton.isEnabled = false
        SVProgressHUD.show()
        
        UserService.instance.signIn(withID: id, token: token, success: { _ in
            let rootViewController = TabBarController()
            UIApplication.shared.keyWindow?.replaceViewController(with: rootViewController)
        }, error: { error in
            var errorMessage: String!
            
            switch error.caronaeCode {
            case .invalidCredentials:
                errorMessage = "Chave não autorizada. Verifique se a mesma foi digitada corretamente e tente de novo."
            case .invalidResponse:
                errorMessage = "Ocorreu um erro carregando seu perfil."
            default:
                errorMessage = "Ocorreu um erro autenticando com o servidor do Caronaê. Por favor, tente novamente."
            }
            
            SVProgressHUD.dismiss()
            CaronaeAlertController.presentOkAlert(withTitle: "Não foi possível autenticar", message: errorMessage)
            self.authButton.isEnabled = true
        })
    }
    
    @objc func openIntranetLogin() {
        let loginURL = URL(string: CaronaeURLString.login)!
        if #available(iOS 9, *) {
            let safariViewController = SFSafariViewController(url: loginURL, entersReaderIfAvailable: false)
            present(safariViewController, animated: true)
        } else {
            UIApplication.shared.openURL(loginURL)
        }
    }

    // MARK: IBActions
    
    @IBAction func didTapAuthenticateButton(_ sender: Any) {
        authenticate(id: idTextField.text!, token: tokenTextField.text!)
    }
    
}

extension TokenViewController: UITextFieldDelegate {
    func textFieldShouldReturn(_ textField: UITextField) -> Bool {
        switch textField {
        case tokenTextField:
            authenticate(id: idTextField.text!, token: tokenTextField.text!)
            return false
        case idTextField:
            if idTextField.hasText {
                tokenTextField.becomeFirstResponder()
            }
        default:
            break
        }
        
        return true
    }
    
    func textField(_ textField: UITextField, shouldChangeCharactersIn range: NSRange, replacementString string: String) -> Bool {
        let text = (textField.text as NSString!).replacingCharacters(in: range, with: string)
        switch textField {
        case idTextField:
            authButton.isEnabled = !text.isEmpty && tokenTextField.hasText
        default:
            authButton.isEnabled = !text.isEmpty && idTextField.hasText
        }
        
        return true
    }
}
